external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/SingleTopicKafkaSpoutTest.java

```diff
@@ -0,0 +1,248 @@
+
+        spout.acked.values().forEach(item -> {
+            assertOffset(messageCount - 1, (KafkaSpout.OffsetEntry) item);
+        });
+    }
+
+    @Test
+    public void shouldEmitAllMessages() throws Exception {
+        int messageCount = 10;
+        populateTopicData(SingleTopicKafkaSpoutConfiguration.topic, messageCount);
+        int kafkaPort = kafkaUnitRule.getKafkaPort();
+
+        TopologyContext topology = mock(TopologyContext.class);
+        SpoutOutputCollector collector = mock(SpoutOutputCollector.class);
+        Map conf = mock(Map.class);
+
+        KafkaSpout spout = new KafkaSpout<>(getKafkaSpoutConfig(getKafkaSpoutStreams(), kafkaPort));
+        spout.open(conf, topology, collector);
+        spout.activate();
+
+        IntStream.range(0, messageCount).forEach(value -> {
+            spout.nextTuple();
+            ArgumentCaptor<Object> messageId = ArgumentCaptor.forClass(Object.class);
+            verify(collector).emit(
+                    eq(SingleTopicKafkaSpoutConfiguration.stream),
+                    eq(new Values(SingleTopicKafkaSpoutConfiguration.topic,
+                            ((Integer)value).toString(),
+                            ((Integer)value).toString())),
+                    messageId.capture());
+            spout.ack(messageId.getValue());
+            reset(collector);
+        });
+
+        spout.acked.values().forEach(item -> {
+            assertOffset(messageCount - 1, (KafkaSpout.OffsetEntry) item);
+        });
+    }
+
+    @Test
+    public void shouldReplayInOrderFailedMessages() throws Exception {
+        int messageCount = 10;
+        populateTopicData(SingleTopicKafkaSpoutConfiguration.topic, messageCount);
+        int kafkaPort = kafkaUnitRule.getKafkaPort();
+
+        TopologyContext topology = mock(TopologyContext.class);
+        SpoutOutputCollector collector = mock(SpoutOutputCollector.class);
+        Map conf = mock(Map.class);
+
+        KafkaSpout spout = new KafkaSpout<>(getKafkaSpoutConfig(getKafkaSpoutStreams(), kafkaPort));
+        spout.open(conf, topology, collector);
+        spout.activate();
+
+        //play and ack 1 tuple
+        ArgumentCaptor<Object> messageIdAcked = ArgumentCaptor.forClass(Object.class);
+        spout.nextTuple();
+        verify(collector).emit(anyObject(), anyObject(), messageIdAcked.capture());
+        spout.ack(messageIdAcked.getValue());
+        reset(collector);
+
+        //play and fail 1 tuple
+        ArgumentCaptor<Object> messageIdFailed = ArgumentCaptor.forClass(Object.class);
+        spout.nextTuple();
+        verify(collector).emit(anyObject(), anyObject(), messageIdFailed.capture());
+        spout.fail(messageIdFailed.getValue());
+        reset(collector);
+
+        //pause so that failed tuples will be retried
+        Thread.sleep(200);
+
+        //allow for some failures with +5 instead of -1
+        IntStream.range(0, messageCount + 5).forEach(value -> {
```

Comment:
A call to the spout's emit() does not always mean a tuple is emitted. I will improve the comment.